### DISCUSSION TITLE
Native marker for solidified code

### DISCUSSION THREAD

**Discussion by s-hadinger** [berry-lang/berry#191] [Ideas]
I have made some improvements in solidify that I need to upload...

Now `solidify` works with modules, classes, lists, maps... But when it comes to classes or modules with native functions, there is no way to mark them as native.

I would like to have a Berry skeleton, solidify it and then add native functions. I see two alternatives to do that.

**1/ add `native` keywork**

``` ruby
class myclass
    native f
    def g() end
end
```

Pros: allow any name
Cons: adds a new reserved keyword

**2/ use a naming convention, like any name starting with `ntv_` or `native_`**

``` ruby
class myclass
    static native_f       # will be recongnized by solidify module that will add a native function stub instead
    def g() end
end
```

Pros: no change in parser
Cons: naming by convention... Some people don't like it

Afterthoughts I prefer option 2.

**Comment by skiars** [comment]
Do you mean that these symbols will generate a placeholder and then implement C native functions by myself?

**Reply by s-hadinger** [reply]
Yes, but in this case the native function already exists. I will provide an example.

**Comment by Sebight** [comment]
Hello, sorry for this necro, however I am wondering whether something like this was ever implemented.

I am facing an issue with this code:

```be
class SpectodaInterface
    def init()
        print('Initializing SpectodaInterface')
    end

    def emit(eventIdentifier, value)
        print("[BE]: Called with eventIdentifier: "+ eventIdentifier + " and value: " + value)
        native_emit(eventIdentifier, value)
    end
end

import solidify
solidify.dump(SpectodaInterface)
```

This will always call this error: `syntax_error: solidify.be:8: 'native_emit' undeclared (first use in this function)`, I even tried doing something like a forward definition:

```be
class SpectodaInterface
    def init()
        print('Initializing SpectodaInterface')
    end

    def native_emit()
        # To be implemented by the native side
    end

    def emit(eventIdentifier, value)
        print("[BE]: Called with eventIdentifier: "+ eventIdentifier + " and value: " + value)
        native_emit(eventIdentifier, value)
    end
end

import solidify
solidify.dump(SpectodaInterface)
```
But still no success.

Note: The native_emit function is register using this `be_regfunc(vm, "native_emit", SpectodaInterface::native_emit);`.

Thank you

**Reply by s-hadinger** [reply]
Generally speaking, `be_regfunc` is not much used and enabled only if `BE_USE_PRECOMPILED_OBJECT` is not defined. Anyways  your function should be declared in the global scope.

**Reply by Sebight** [reply]
So should I look into berry-mapping instead? And also when I put it into the global scope, there won't be a problem with the class solidify? I'll get to it on Monday.

**Reply by s-hadinger** [reply]
No sorry, I wasn't clear, there are several ways to do it.

My preferred one is:
- create a native C class `SpectodaInterface_ntv` using the precompiled syntax in C, and declare here all C code
- create a Berry `class SpectodaInterface : SpectodaInterface_ntv` that inherits from the native class, and that can be solidified

**Reply by Sebight** [reply]
Hi, sorry if I'm too confused. What exactly do you mean with the "precompiled syntax"? Could I find an example somewhere, maybe the docs or Tasmota? Thanks for help

**Reply by s-hadinger** [reply]
Sure let's take it step by step.

First define `be_SpectodaInterface_ntv.c`:

```C
#include "be_constobj.h"

#include "be_fixed_be_class_SpectodaInterface_ntv.h" // this header file is generated by coc compiler

extern int s_native_emit(bvm *vm);

/* @const_object_info_begin

class be_class_SpectodaInterface_ntv (scope: global, name: SpectodaInterface_ntv) {
    native_emit, static_func(s_native_emit)
}
@const_object_info_end */
```

You can then define your function in a C++ file:
```C
extern "C" int s_native_emit(bvm *vm) {
  // do your stuff
}
```

Then add the native class definition in `be_modtab.c`:
```C
[...]
be_extern_native_class(SpectodaInterface_ntv);


BERRY_LOCAL bclass_array be_class_table = {
    /* first list are direct classes */
    /* &be_native_class(my_class), */
    &be_native_class(SpectodaInterface_ntv),
    NULL, /* do not remove */
};
```

Compile Berry, and test that you can access `SpectodaInterface_ntv.native_emit`.

You can then define a Berry file with:
```berry
class SpectodaInterface : SpectodaInterface_ntv
    def init()
        print('Initializing SpectodaInterface')
    end

    def emit(eventIdentifier, value)
        print("[BE]: Called with eventIdentifier: "+ eventIdentifier + " and value: " + value)
        self.native_emit(eventIdentifier, value)
    end
end

import solidify
solidify.dump(SpectodaInterface)
```

**Reply by Sebight** [reply]
Hi! Thanks for the thorough explanation.

I have created this be_SpectodaInterface_ntv.h file: 
```cpp
#include "be_constobj.h"
#include "../../berry/generate/be_fixed_be_class_SpectodaInterface_ntv.h"

extern int s_native_emit(bvm* vm);

/* @const_object_info_begin

class be_class_SpectodaInterface_ntv (scope: global, name: SpectodaInterface_ntv) {
    native_emit, static_func(s_native_emit)
}
@const_object_info_end */
```

After running the coc, this file is created:
```h
#include "be_constobj.h"

static be_define_const_map_slots(be_class_SpectodaInterface_ntv_map) {
    { be_const_key(native_emit, -1), be_const_static_func(s_native_emit) },
};

static be_define_const_map(
    be_class_SpectodaInterface_ntv_map,
    1
);

BE_EXPORT_VARIABLE be_define_const_class(
    be_class_SpectodaInterface_ntv,
    0,
    NULL,
    SpectodaInterface_ntv
);
```
When I try to compile the project I get an error saying:
![image](https://github.com/berry-lang/berry/assets/59232422/91a2db73-7148-438a-a2e7-4d8051391af7)
that refers to this line `{ be_const_key(native_emit, -1), be_const_static_func(s_native_emit) },`

This reminds me of the keyword problems that I used to have earlier, and those were usually generated in `be_const_strtab_def.h`, but nothing is there now. How should I fix this?

Thanks a lot.

**Reply by s-hadinger** [reply]
You're almost there. You need to include the generated headers after the extern functions definitions:

```C
extern int s_native_emit(bvm* vm);

#include "../../berry/generate/be_fixed_be_class_SpectodaInterface_ntv.h"
```
in that order

My bad, I put them in the wrong order up there.

**Reply by Sebight** [reply]
(Haha, I just noticed your reply 😄, thank you)

Turns out I had to put at the end (or at least behind the extern statement):
```c
#include "be_constobj.h"

extern int s_native_emit(bvm* vm);

/* @const_object_info_begin

class be_class_SpectodaInterface_ntv (scope: global, name: SpectodaInterface_ntv) {
    native_emit, static_func(s_native_emit)
}
@const_object_info_end */
#include "../../berry/generate/be_fixed_be_class_SpectodaInterface_ntv.h"
```

With this done and compiling, I did this in `be_modtab.c` (honestly didn't know whether the one in port or in the berry/generate, so I did it in both):
```c
/* user-defined classes declare start */
/* be_extern_native_class(my_class); */
be_extern_native_class(SpectodaInterface_ntv);
/* user-defined classes declare end */

BERRY_LOCAL bclass_array be_class_table = {
    /* first list are direct classes */
    /* &be_native_class(my_class), */
    &be_native_class(SpectodaInterface_ntv),
    NULL, /* do not remove */
};
```

However, when I call this `solidify.be` script:
```be
class SpectodaInterface : SpectodaInterface_ntv
    def init()
        print('Initializing SpectodaInterface')
    end

    def emit(eventIdentifier, value)
        print("[BE]: Called with eventIdentifier: "+ eventIdentifier + " and value: " + value)
        self.native_emit(eventIdentifier, value)
    end
end

import solidify
solidify.dump(SpectodaInterface)
```

Using this command:
`components\berry\berry.exe -s -g solidify.be` 

I get this error: `syntax_error: solidify.be:1: 'SpectodaInterface_ntv' undeclared (first use in this function)`.

Does this have to do something with the modtab.c file not being correctly acknowledged of my changes in terms of the new native class?

Thanks

**Reply by s-hadinger** [reply]
This is weird. If the change is done in modtab.c then the class should be visible. Are you using the `berry.exe` that was compiled with the changed modtab.c ?

Edit: are you doing cross-platform compilation?

**Reply by Sebight** [reply]
I am using the only berry.exe in the project for the solidification process, something we want done in the "editor" before flashing the ESP32.

I just tried to solidify it on the ESP32 and that seems to work (or at least it spat out the C code, I will test it later), so it might be something with the berry.exe being outdated, I'll look around to see what's up.

**Reply by s-hadinger** [reply]
In such case, you just need to make sure that `SpectodaInterface_ntv` exists in the global scope when `solidify.be` is ran. Are you using the one from Tasmota?

Or just `SpectodaInterface_ntv = nil` should be enough to have solidification work successfully

**Reply by Sebight** [reply]
It loads a wrong build, as can be seen here:
![image](https://github.com/berry-lang/berry/assets/59232422/5bd0c413-488b-44af-97bc-1572b5100b69)

Once I call `make` (that regens the .exe file, as I understood), it fails at a linker problem: 
![image](https://github.com/berry-lang/berry/assets/59232422/fe9fb199-5aae-4dbf-b61e-51d097466e44)

My solidify.be is as simple as this:
```be
class SpectodaInterface : SpectodaInterface_ntv
    def init()
        print('Initializing SpectodaInterface')
    end

    def emit(eventIdentifier, value)
        print("[BE]: Called with eventIdentifier: "+ eventIdentifier + " and value: " + value)
        self.native_emit(eventIdentifier, value)
    end
end

import solidify
solidify.dump(SpectodaInterface)
```

being called from this Python script (hardcoded for the usecase right now): 
```python
import os
import sys
import subprocess

IS_WINDOWS = sys.platform.startswith("win")
BERRY_PATH = "path"
FILE_NAME = "solidify.be"

def getBerry(): 
    if IS_WINDOWS:
        return BERRY_PATH + ".exe"


solidify_command = getBerry()
solidify_flags = " -s -g " + FILE_NAME
print(solidify_command + solidify_flags)
print("#include \"be_constobj.h\"")
subprocess.call(solidify_command + solidify_flags, shell=True)
```

I would test the `SpectodaInterface_ntv = nil`, but my .exe is now gone. :D So I need to get it to compile.

EDIT:
Tested this:
```be
SpectodaInterface_ntv = nil
class SpectodaInterface : SpectodaInterface_ntv
    def init()
        print('Initializing SpectodaInterface')
    end

    def emit(eventIdentifier, value)
        print("[BE]: Called with eventIdentifier: "+ eventIdentifier + " and value: " + value)
        self.native_emit(eventIdentifier, value)
    end
end

import solidify
solidify.dump(SpectodaInterface)
```
with error: `type_error: value 'nil' does not support set super stack traceback: solidify.be:2: in function main`

**Reply by s-hadinger** [reply]
To be clear, the modtab.c file used or your laptop for solidification must not have the class definition. It's only needed on the target device.

Maybe `nil` is not enough. Try `class SpectodaInterface_ntv end` instead of `SpectodaInterface_ntv = nil`

**Reply by Sebight** [reply]
Okay, thank you. Things are looking _a bit_ better now.

solidify.be:
```be
class SpectodaInterface_ntv end
class SpectodaInterface : SpectodaInterface_ntv
    def init()
        print('Initializing SpectodaInterface')
    end

    def emit(first, second, third)
        print("[BE] #1: " + str(first))
        print("[BE] #2: " + str(second))
        print("[BE] #3: " + str(third))
        print("[BE] #4: " + str(self))
        # print("[BE]: Called with eventIdentifier: "+ eventIdentifier + " and value: " + value)
        # self.native_emit(eventIdentifier, value)
        self.native_emit(first, second)
    end
end

import solidify
solidify.dump(SpectodaInterface)
```
This gets solidified on PC just fine.

This is my main Berry code:
```be
def start() end

def handle(context)
    # if (!initialized) 
    #     return
    # end
    print(spectoda)
    print(SpectodaInterface_ntv)
    print(spectoda.emit)
    spectoda.emit("test", "abc")
end
```
It runs, up to one point, this is the output:
```
<class: SpectodaInterface>
<class: SpectodaInterface_ntv>
<function: 0x3f400668>
[BE] #1: abc
[BE] #2: nil
[BE] #3: nil
[BE] #4: test
E (439) main: Failed to call handle: 'string' value has no method 'native_emit'
```

First three logs are fine, as expected.

When it comes to the first parameter log, why is it taking the second parameter and `self` is then "test" aka the first parameter.

Is there a problem with the stack order or something?

Thank you

**Reply by s-hadinger** [reply]
It's hard to tell because I don't know what `native_emit` actually does.

You initially defined `native_emit` as a global function (not a member class), that's why I have indicated to declare it as `static_func(s_native_emit)` which means that it does not take `self` as a first parameter. Now if you need `self`, declare it as `func(s_native_emit)` instead

**Reply by Sebight** [reply]
Yes, I am sorry. I'm still quite confused about what I want to do, as I'm still learning about how Berry works.

My final usecase is this, separated in three layers:
- Native C++ classes that I want to "inject" into Berry as globals. These will be directly linked to C++ and are classes such as `spectoda`, `gpio`, `...`
- Berry drivers that will get solidified and will interact with these native globals, such as `BerryExampleDriver`
- End-user Berry code that will have a few life-cycle functions called directly from C++ (start, update, ...) and will be used to interact with the Berry drivers => `BerryExampleDriver.calculateAThing()`

I have a few questions though:
- Should I keep the native classes static to represent some sort of a singleton? I suppose that could work.
- Those solidified Berry drivers should, however, be normal classes that can get instantiated. Or do you think otherwise?

I'm certain you have much more experience in terms of building architecture around Berry, so would you mind sharing your thoughts? Thanks a lot!

**Reply by s-hadinger** [reply]
Sounds awesome.

There is no definite pattern for singletons, this is what I do:
- `Tasmota_ntv` is a native class with C/C++ code. I have a mix of static (class methods) and regular methods
- `Tasmota` is a Berry sub-class of `Tasmota_ntv` and is solidified
- `tasmota` is a singleton instance of this class that is created in a startup script

This pattern is straightforward. I haven't found any simpler or better way until now.

**Reply by Sebight** [reply]
Looks sweet! However, why do you create the Berry layer for your Tasmota class? Couldn't you just make Tasmota_ntv as "tasmota"

In _something_ like this:
```
class be_class_Tasmota_ntv (scope: global, name: tasmota) 
```

Because for me, my `tasmota` class will be just directly calling C++ functions, no additional Berry implementation.

**Reply by s-hadinger** [reply]
Because I need to keep some data around. For example I keep a list of custom "drivers" in a list as an instance variable of the singleton; this avoids polluting the global namespace.

Edit: I should add that a solidified class has its static variables read-only
